segmentstore/server/host/src/main/java/io/pravega/segmentstore/server/host/stat/SegmentAggregates.java

```diff
@@ -156,25 +164,95 @@ private double decayingRate(long count, double rate, double alpha, double interv
         }
     }
 
-    double getTwoMinuteRate() {
-        return Double.longBitsToDouble(twoMinuteRate.get());
+    synchronized int getTargetRate() {
+        return targetRate;
     }
 
-    double getFiveMinuteRate() {
-        return Double.longBitsToDouble(fiveMinuteRate.get());
+    synchronized void setTargetRate(int value) {
+        targetRate = value;
     }
 
-    double getTenMinuteRate() {
-        return Double.longBitsToDouble(tenMinuteRate.get());
+    synchronized double getTwoMinuteRate() {
+        return twoMinuteRate;
     }
 
-    double getTwentyMinuteRate() {
-        return Double.longBitsToDouble(twentyMinuteRate.get());
+    synchronized double getFiveMinuteRate() {
+        return fiveMinuteRate;
     }
 
-    @VisibleForTesting
-    static void setClock(Clock clock) {
-        CLOCK.set(clock);
+    synchronized double getTenMinuteRate() {
+        return tenMinuteRate;
+    }
+
+    synchronized double getTwentyMinuteRate() {
+        return twentyMinuteRate;
+    }
+
+    void reportIfNecessary(Duration reportingDuration, Runnable callback) {
+        lastReportedTime.getAndUpdate(prev -> {
+            if (getTimeMillis() - prev > reportingDuration.toMillis()) {
+                callback.run();
+                return getTimeMillis();
```

Comment:
I thought about it too. I think the goal here is to store the time when the last report ended. `callback.run()` may take a while to execute so, based on the existing behavior, I assumed this wants to keep a delay between the ending of the previous run and the beginning of the next. 

Otherwise, if `callback.run` takes too long, I could see this trying to overlap executions and end up running continuously.

Reply:
Right, but think about what happens in the event that race occurs now:
Thread 1 calls getAndUpdate it get's the time and runs the callback.
Thread 2 does the same thing and runs to completion changing the time upon its completion.
Thread 1's update fails, and so thread 1 RE-RUNS the callback.
Surely this is undesirable behavior. There isn't even an upper bound on how many times reportIfNecessary will be invoked. If the callback takes longer than `reportingDuration` to run, this could be an infinite loop.